### PR TITLE
feat(c/sedona-geos): Implement ST_IsRing

### PR DESCRIPTION
**Pull request by joonaspessi**
Hello!

Thought that implementing something relatively simple from your [first issue proposal](https://github.com/apache/sedona-db/issues/224) would be a good way to understand how `sedona-db` works. Thanks for creating this project.

This PR implements `ST_IsRing` with GEOS. 

I did some research on how other spatial libraries handle this function when a non-LineString geometry is passed:

  | System | Non-LineString Behavior |
  |--------|------------------------|
  | **DuckDB spatial extensions** | Returns `false` |
  | **Apache Sedona** | Returns `null` |
  | **PostGIS** | Throws error  |
  | **This PR** | Returns `false`|

  I chose to follow the original Apache Sedona rather than PostGIS because it matches the existing Sedona ecosystem.

### CODE REVIEW

File: c/sedona-geos/src/st_isring.rs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    let is_closed = geos_geom.is_closed().map_err(|e| {
+        DataFusionError::Execution(format!("Failed to check if geometry is closed: {e}"))
+    })?;
+
+    // Check if the LineString is simple (no self-intersections)
+    let is_simple = geos_geom.is_simple().map_err(|e| {
+        DataFusionError::Execution(format!("Failed to check if geometry is simple: {e}"))
+    })?;
+
+    Ok(is_closed && is_simple)
```
Comment (petern48) [inline c/sedona-geos/src/st_isring.rs]:
```suggestion
    Ok(geos_geom.is_ring().map_err(|e| {
        DataFusionError::Execution(format!("Failed to check if geometry is a ring: {e}"))
    })?)
```

We can actually simplify this code to simply call `geos`'s `.is_ring()` method [here](https://github.com/georust/geos/blob/47afbad2483e489911ddb456417808340e9342c3/src/geometry.rs#L252-L264). Seems to pass everything when I tried this locally.

File: python/sedonadb/tests/functions/test_functions.py [hunk 2]
```diff
@@ -604,6 +604,52 @@ def test_st_isclosed(eng, geom, expected):
...
+        ("LINESTRING(0 0, 1 0, 1 1, 0 0)", True),
+        ("LINESTRING(0 0, 2 2, 1 2, 0 0)", True),
+        # Closed but self-intersecting - bowtie shape
+        ("LINESTRING(0 0, 0 1, 1 0, 1 1, 0 0)", False),
+        # Not closed
+        ("LINESTRING(0 0, 1 1)", False),
+        ("LINESTRING(2 0, 2 2, 3 3)", False),
+        ("LINESTRING(0 0, 2 2)", False),
+        # Empty LineString
+        ("LINESTRING EMPTY", False),
```
Comment (petern48) [inline python/sedonadb/tests/functions/test_functions.py:680]:
```suggestion
        ("LINESTRING EMPTY", False),
        # Collections of linestrings that would have been considered rings
        ("MULTILINESTRING((0 0, 0 1, 1 1, 1 0, 0 0))", False),
        ("GEOMETRYCOLLECTION(LINESTRING(0 0, 0 1, 1 1, 1 0, 0 0))", False),
```

Let's add these two cases. Returning `False` is the natural behavior for the `geos`'s `is_ring()` method, which agrees with DuckDB

**Review by petern48 — COMMENTED** [review]
Welcome! I left a few comments, but it looks good so far. Just need to agree on which behavior to follow.

**Comment by petern48** [comment]
I personally agree with the proposed behavior. [PostGIS's implementation](https://postgis.net/docs/doxygen/3.5/d4/d06/postgis_2lwgeom__geos_8c_aee8112fca59e61cf6c5938ba03096865.html) is odd imo. For non-linestrings, it returns `ERROR:  ST_IsRing() should only be called on a linear feature` if it's non-empty, but returns False if it's empty. The thing I don't like about this is that if you have **any** non-empty non-linestring geometry in your column, your query will just fail.

```sql
CREATE TABLE geom (geom GEOMETRY);
INSERT INTO geom VALUES (ST_GeomFromText('LINESTRING EMPTY')), (ST_Point(3, 4));
SELECT st_isring(geom) FROM geom;
-- ERROR:  ST_IsRing() should only be called on a linear feature
```

DuckDB returns `False` instead of the error, which agrees with the natural behavior of the `is_ring()` function. Sedona actually returns `NULL` for any non-linestring geometry (empty or not).

WDYT @paleolimbot?

**Review by paleolimbot — COMMENTED** [review]
Thank you!

> DuckDB returns False instead of the error, which agrees with the natural behavior of the is_ring() function. Sedona actually returns NULL for any non-linestring geometry (empty or not).

I think we should follow PostGIS here and file a bug in Sedona-spark to discuss if that decision was intentional. If a user is trying to check the ring status of something that isn't even the correct geometry type there is likely an error in the query or an error in the data.

In the next DataFusion release we'll have the ability to pipe options into scalar functions and so we can do things like enable a Sedona-spark compatibility mode. Here I think PostGIS compatibility is the right choice because it's the easiest to test and review (and in general has more intentional corner case behaviour.

**Comment by jiayuasu** [comment]
Fine by me. PostGIS compatibility is more important

File: python/sedonadb/tests/functions/test_functions.py (hunk 2 above)
Comment (joonaspessi) [reply python/sedonadb/tests/functions/test_functions.py:680]: Makes sense. Added these under `test_st_isring_non_linestring` as PostGis is throwing error on these.

**Comment by joonaspessi** [comment]
Thanks for the comments. I think the PostGis compatibility is a good choice to follow. Updated the PR to follow that approach. While doing it, also noticed that for empty geometries, PostGis will return `false`.

File: python/sedonadb/tests/functions/test_functions.py [hunk 3]
```diff
@@ -661,6 +661,66 @@ def test_st_isclosed(eng, geom, expected):
...
+    """
+    eng = eng.create_or_skip()
+    with pytest.raises(Exception) as exc_info:
+        eng.assert_query_result(f"SELECT ST_IsRing(ST_GeomFromText('{geom}'))", None)
+
+    # Verify error message contains the expected text
+    error_msg = str(exc_info.value).lower()
+    assert (
+        "linear" in error_msg or "linestring" in error_msg
+    ), f"Expected error about linear feature, got: {exc_info.value}"
```
Comment (paleolimbot) [inline python/sedonadb/tests/functions/test_functions.py]: I think this will work:

```suggestion
    with pytest.raises(Exception, match="linear|linestring") as exc_info:
        eng.assert_query_result(f"SELECT ST_IsRing(ST_GeomFromText('{geom}'))", None)
```

Comment (joonaspessi) [reply python/sedonadb/tests/functions/test_functions.py]: Good idea! Updated to cleanly utilize pytest match pattern. Pre-commit passes locally

**Review by paleolimbot — APPROVED** [review]
Thank you!
